### ISSUE TITLE
Rework cloud providers to lookup by instanceID not nodeName

### ISSUE THREAD

**Issue by sjenning** [kubernetes/kubernetes#31646]
I've been looking into how to handle a issue surrounding the Openstack provider code assuming that the node name is equal to the instance name.  It seems like much of the cloud providers use searches by name in `ExternalID()` (deprecated) and `InstanceID()` to convert the node name to a cloud instanceID.

Currently, afaict, these are only used in a few places:
- Load balancer attach (provider internal)
- Volume attach (provider internal)
- Kubelet node resource creation
- Node Controller to remove nodes that are no longer in the cloud

It seems to me that the decoupling the node name completely from the cloud provider lookups would be a good thing, using the ExternalID field in the node resource to contain the cloud-specific instance ID to establish the node-to-instance mapping rather than doing a lookup by node name, which creates a layering dependency (the aforementioned assumptions about the node name meaning something at the cloud layer).

What do people think about this?  I don't mind doing the work, just don't want to go off on a fool's errand if there is a reason this is not done already.

@kubernetes/rh-cluster-infra @anguslees @mikedanese

**Comment by anguslees** [comment]
My approach with https://github.com/kubernetes/kubernetes/pull/31321 (awaiting review) is to pass nodes as full `Node` objects, rather than using names or IDs directly.  That way the provider code can just use the attributes it wants (including reading `NodeAddresses` directly, in the cases where we just want to go node -> address).  The instance ID (and addresses) are of course filled out locally on the kubelet via openstack metadata (and soon config-drive) lookups, and then republished around to the rest of the k8s control jobs via `Node.Status` - so we should be able to avoid ever looking at the instance name (yay).

PR 31321 only fixes up the LoadBalancer API.  I haven't gone looking for other places yet, but the same approach should work fine afaik.

The only reason we really have instance names used anywhere is for the ancient `Instance.List` API.  Back in the olden days, we used to specify nodes by a regex on names understood by the provider.  A bunch of these decisions haven't made sense since we shifted to adding nodes explicitly.  I think we should start the above plan by formally deleting the `Instances.List` API and it's implied ability to string-match usefully.

So: Yes to your plan, only I'd suggest going further and using full `Node` objects wherever possible in the k8s API to refer to nodes.  Once we have the full object available within the (openstack) provider, we should indeed move to using `ExternalID` directly wherever possible.
